pgx/src/datum/sql_entity_graph/mod.rs

```diff
@@ -64,6 +64,72 @@ pub trait ToSql {
     fn to_sql(&self, context: &PgxSql) -> eyre::Result<String>;
 }
 
+/// Represents a pointer to a function that can transform a SqlGraphEntity to a SQL string
+///
+/// This is used to provide a facility for overriding the default SQL generator behavior using
+/// the `#[to_sql(path::to::function)]` attribute in circumstances where the default behavior is
+/// not desirable.
+///
+/// Implementations can invoke `ToSql::to_sql(entity, context)` on the unwrapped SqlGraphEntity
+/// type should they wish to delegate to the default behavior for any reason.
+pub type ToSqlFn = fn(&SqlGraphEntity, &PgxSql) -> eyre::Result<String>;
```

Comment:
I don't really want to infect downstream users with `eyre::Result`. Can we do a `Result<String, impl std::error::Error>` here?

Reply:
I'll experiment a bit and see, but I'm sure we can use something along those lines, as long as it is convertible to `eyre::Result` internally. For `eyre::Result` that should be anything that implements `std::error::Error + Send + Sync + 'static`

Reply:
Yeah, that'd be good. That means that users won't necessarily need to use eyre. :)

Reply:
Eyre guides us here! https://docs.rs/eyre/latest/eyre/trait.WrapErr.html#wrapping-types-that-dont-impl-error-eg-str-and-boxdyn-error

This works:

```patch
diff --git a/pgx-tests/src/tests/schema_tests.rs b/pgx-tests/src/tests/schema_tests.rs
index 8c16bc2..0953079 100644
--- a/pgx-tests/src/tests/schema_tests.rs
+++ b/pgx-tests/src/tests/schema_tests.rs
@@ -30,7 +30,7 @@ mod test_schema {
     #[to_sql(generate_type)]
     pub struct OtherType(pub u64);
 
-    fn generate_function(entity: &SqlGraphEntity, _context: &PgxSql) -> eyre::Result<String> {
+    fn generate_function(entity: &SqlGraphEntity, _context: &PgxSql) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
         if let SqlGraphEntity::Function(ref func) = entity {
             Ok(format!("\
                 CREATE OR REPLACE FUNCTION test_schema.\"func_generated_with_custom_name\"() RETURNS void\n\
@@ -44,7 +44,7 @@ mod test_schema {
         }
     }
 
-    fn generate_type(entity: &SqlGraphEntity, _context: &PgxSql) -> eyre::Result<String> {
+    fn generate_type(entity: &SqlGraphEntity, _context: &PgxSql) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
         if let SqlGraphEntity::Type(ref ty) = entity {
             Ok(format!(
                 "\n\
diff --git a/pgx/src/datum/sql_entity_graph/mod.rs b/pgx/src/datum/sql_entity_graph/mod.rs
index e86dbf2..d41f0a8 100644
--- a/pgx/src/datum/sql_entity_graph/mod.rs
+++ b/pgx/src/datum/sql_entity_graph/mod.rs
@@ -72,7 +72,7 @@ pub trait ToSql {
 ///
 /// Implementations can invoke `ToSql::to_sql(entity, context)` on the unwrapped SqlGraphEntity
 /// type should they wish to delegate to the default behavior for any reason.
-pub type ToSqlFn = fn(&SqlGraphEntity, &PgxSql) -> eyre::Result<String>;
+pub type ToSqlFn = fn(&SqlGraphEntity, &PgxSql) -> Result<String, Box<dyn std::error::Error + Send + Sync + 'static>>;
 
 /// Represents configuration options for tuning the SQL generator.
 ///
diff --git a/pgx/src/datum/sql_entity_graph/sql_graph_entity.rs b/pgx/src/datum/sql_entity_graph/sql_graph_entity.rs
index 9ea96ab..d85076e 100644
--- a/pgx/src/datum/sql_entity_graph/sql_graph_entity.rs
+++ b/pgx/src/datum/sql_entity_graph/sql_graph_entity.rs
@@ -3,6 +3,7 @@ use super::{
     PostgresEnumEntity, PostgresHashEntity, PostgresOrdEntity, PostgresTypeEntity, SchemaEntity,
     SqlGraphIdentifier, ToSql,
 };
+use eyre::{WrapErr, eyre};
 
 /// An entity corresponding to some SQL required by the extension.
 #[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
@@ -94,7 +95,7 @@ impl ToSql for SqlGraphEntity {
             },
             SqlGraphEntity::Function(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Function(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else if context.graph.neighbors_undirected(context.externs.get(item).unwrap().clone()).any(|neighbor| {
                 let neighbor_item = &context.graph[neighbor];
                 match neighbor_item {
@@ -118,32 +119,32 @@ impl ToSql for SqlGraphEntity {
             }
             SqlGraphEntity::Type(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Type(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else {
                 item.to_sql(context)
             }
             SqlGraphEntity::BuiltinType(_) => Ok(String::default()),
             SqlGraphEntity::Enum(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Enum(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else {
                 item.to_sql(context)
             }
             SqlGraphEntity::Ord(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Ord(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else {
                 item.to_sql(context)
             }
             SqlGraphEntity::Hash(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Hash(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else {
                 item.to_sql(context)
             }
             SqlGraphEntity::Aggregate(item) if !item.to_sql_config.enabled => Ok(String::default()),
             SqlGraphEntity::Aggregate(item) => if let Some(callback) = item.to_sql_config.callback {
-                callback(self, context)
+                callback(self, context).map_err(|e| eyre!(e)).wrap_err("Failed to run specified `#[pgx(sql = path)] function")
             } else {
                 item.to_sql(context)
             }
```

Reply:
Ah nice! I thought I tried that, but `WrapErr` was the missing piece of the puzzle :)